<a href="https://colab.research.google.com/github/NaufalYogaPratama/learning-data-science/blob/main/ml-basics/unsupervised-learning-clustering/clustering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Clustering Algorithm

## Mengimport Semua Library

In [1]:
import pandas as pd  # Mengimpor pustaka pandas untuk manipulasi dan analisis data
import matplotlib.pyplot as plt  # Mengimpor pustaka matplotlib untuk visualisasi grafik
from yellowbrick.cluster import KElbowVisualizer  # Mengimpor KElbowVisualizer untuk visualisasi metode Elbow

from sklearn.cluster import KMeans, DBSCAN  # Mengimpor algoritma KMeans dan DBSCAN untuk clustering
from sklearn.metrics import silhouette_score  # Mengimpor silhouette_score untuk mengevaluasi hasil clustering

## Data Loading

In [2]:
# Membaca dataset pelanggan mall dari URL dan menampilkan 5 baris pertama
df = pd.read_csv('https://raw.githubusercontent.com/dicodingacademy/dicoding_dataset/main/ML%20Pemula/Mall_Customers.csv')
df.head()

CustomerID  Gender  Age  Annual Income (k$)  Spending Score (1-100)
0           1    Male   19                  15                      39
1           2    Male   21                  15                      81
2           3  Female   20                  16                       6
3           4  Female   23                  16                      77
4           5  Female   31                  17                      40

In [3]:
# Menampilkan informasi tentang dataset, termasuk jumlah baris, kolom, tipe data, dan jumlah nilai non-null
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   CustomerID              200 non-null    int64 
 1   Gender                  200 non-null    object
 2   Age                     200 non-null    int64 
 3   Annual Income (k$)      200 non-null    int64 
 4   Spending Score (1-100)  200 non-null    int64 
dtypes: int64(4), object(1)
memory usage: 7.9+ KB


### Dataset Overview
- **Jumlah Baris**: 200
- **Jumlah Kolom**: 5

---

### Deskripsi Kolom

1. **CustomerID**  
   - **Tipe Data**: `int64`  
   - **Deskripsi**: ID unik untuk mengidentifikasi setiap pelanggan.

2. **Gender**  
   - **Tipe Data**: `object` (kategori)  
   - **Deskripsi**: Jenis kelamin pelanggan (contoh: Male/Female).

3. **Age**  
   - **Tipe Data**: `int64`  
   - **Deskripsi**: Usia pelanggan dalam tahun.

4. **Annual Income (k$)**  
   - **Tipe Data**: `int64`  
   - **Deskripsi**: Pendapatan tahunan pelanggan dalam satuan ribuan dolar (contoh: 15 = \$15.000).

5. **Spending Score (1-100)**  
   - **Tipe Data**: `int64`  
   - **Deskripsi**: Skor pengeluaran pelanggan yang dinilai dari 1 hingga 100.

In [4]:
# Menampilkan statistik deskriptif dari dataset untuk kolom numerik
df.describe()

CustomerID         Age  Annual Income (k$)  Spending Score (1-100)
count  200.000000  200.000000          200.000000              200.000000
mean   100.500000   38.850000           60.560000               50.200000
std     57.879185   13.969007           26.264721               25.823522
min      1.000000   18.000000           15.000000                1.000000
25%     50.750000   28.750000           41.500000               34.750000
50%    100.500000   36.000000           61.500000               50.000000
75%    150.250000   49.000000           78.000000               73.000000
max    200.000000   70.000000          137.000000               99.000000

### Statistik Deskriptif Kolom Numerik

#### **CustomerID**
- **Jumlah Data Unik**: `200`  
- **Rata-Rata**: `100.5`  
- **Rentang Nilai**: `1 - 200`

---

#### **Age (Usia)**  
- **Rentang Usia**: `18 - 70 tahun`  
- **Rata-Rata Usia**: `38.85 tahun`  
- **Distribusi Kuartil**:  
  - **Q1 (25%)**: `28.75 tahun`  
  - **Q3 (75%)**: `49 tahun`  

---

#### **Annual Income (k$)**  
- **Rentang Pendapatan**: `15 - 137 ribu dolar`  
- **Rata-Rata Pendapatan**: `60.56 ribu dolar`  
- **Distribusi Kuartil**:  
  - **Q1 (25%)**: `41.5 ribu dolar`  
  - **Q3 (75%)**: `78 ribu dolar`  

---

#### **Spending Score (1–100)**  
- **Rentang Skor**: `1 - 99`  
- **Rata-Rata Skor**: `50.2`  
- **Distribusi Kuartil**:  
  - **Q1 (25%)**: `34.75`  
  - **Q3 (75%)**: `73`